go/multipooler/executor/executor.go

```diff
@@ -180,169 +186,265 @@ func (e *Executor) PortalStreamExecute(
+		"shard", target.Shard,
+		"user", user,
+		"has_statement", preparedStatement != nil,
+		"has_portal", portal != nil)
 
-		resultRows = append(resultRows, &query.Row{Values: values})
-		rowCount++
+	// Get a connection from the pool
+	conn, err := e.poolManager.GetRegularConnWithSettings(ctx, settings, user)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connection for user %s: %w", user, err)
 	}
+	defer conn.Recycle()
 
-	if err := rows.Err(); err != nil {
-		return nil, fmt.Errorf("error reading rows: %w", err)
+	// Describe prepared statement
+	// Ensure the statement is prepared on this connection
+	canonicalName, err := e.ensurePrepared(ctx, conn.Conn, preparedStatement)
+	if err != nil {
+		return nil, err
 	}
 
-	// Generate command tag for SELECT
-	commandTag := fmt.Sprintf("SELECT %d", rowCount)
+	// If portal is provided, we need to bind first, then describe
+	if portal != nil {
+		// Bind and describe using canonical name
+		paramFormats := int32ToInt16Slice(portal.ParamFormats)
+		resultFormats := int32ToInt16Slice(portal.ResultFormats)
+		desc, err := conn.Conn.BindAndDescribe(ctx, canonicalName, portal.Params, paramFormats, resultFormats)
+		if err != nil {
+			return nil, fmt.Errorf("failed to describe portal: %w", err)
+		}
+		return desc, nil
+	}
 
-	return &query.QueryResult{
-		Fields:       fields,
-		RowsAffected: 0, // SELECT queries don't affect rows
-		Rows:         resultRows,
-		CommandTag:   commandTag,
-	}, nil
+	// Describe prepared using canonical name
+	desc, err := conn.Conn.DescribePrepared(ctx, canonicalName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe prepared statement: %w", err)
+	}
+	return desc, nil
 }
 
-// executeModifyQuery executes an INSERT, UPDATE, DELETE, or other modification query.
-func (e *Executor) executeModifyQuery(ctx context.Context, queryStr string) (*query.QueryResult, error) {
-	result, err := e.db.ExecContext(ctx, queryStr)
-	if err != nil {
-		return nil, fmt.Errorf("failed to execute query: %w", err)
+// ensurePrepared ensures the prepared statement is available on the connection.
+// It uses the consolidator to get a canonical statement name and checks the connection state
+// to avoid redundant parsing. Returns the canonical statement name to use.
+func (e *Executor) ensurePrepared(ctx context.Context, conn *regular.Conn, stmt *query.PreparedStatement) (string, error) {
+	// We use connId 0 since we just need the canonical name mapping
+	// Get the canonical prepared statement info
+	psi := e.consolidator.GetPreparedStatementInfo(0, stmt.Name)
+	if psi == nil {
+		// Add to consolidator to get/create canonical name
+		if err := e.consolidator.AddPreparedStatement(0, stmt.Name, stmt.Query, stmt.ParamTypes); err != nil {
+			return "", fmt.Errorf("failed to consolidate prepared statement: %w", err)
+		}
+		// This will be not-nil because we just added it.
+		psi = e.consolidator.GetPreparedStatementInfo(0, stmt.Name)
```

Comment:
Would it make sense for AddPreparedStatement() to return the statement so it's not necessary to call Get() immediately afterwards and assume it's non-nil?

Reply:
Fixed